### PR TITLE
Adding status update events

### PR DESCRIPTION
**Pull request by vpratfr**
Hi Spatie!

As promised, here is a PR for firing events when status gets updated.

I have included 2 tests for that new feature too.

I also took this opportunity to add some PHPDoc to the trait methods and refactor setStatus to use forceSetStatus in order not to duplicate code.

### CODE REVIEW

File: src/Events/StatusUpdated.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    private $oldStatus;
+
+    /** @var string */
+    private $newStatus;
+
+    /** @var \Illuminate\Database\Eloquent\Model */
+    private $model;
+
+    /**
+     * StatusUpdated constructor.
```
Comment (freekmurze) [inline src/Events/StatusUpdated.php]: Remove this docblock, it doesn't contain any useful info.

File: src/Events/StatusUpdated.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Event fired by the HasStatus trait when a status is updated
+ *
+ * @package Spatie\ModelStatus\Events
+ */
+class StatusUpdated
+{
+    /** @var string */
+    private $oldStatus;
```
Comment (freekmurze) [inline src/Events/StatusUpdated.php]: Make all these properties `protected`

File: src/Events/StatusUpdated.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public function __construct(string $oldStatus, string $newStatus, Model $model)
+    {
+        $this->oldStatus = $oldStatus;
+        $this->newStatus = $newStatus;
+        $this->model = $model;
+    }
+
+    /**
+     * @return string
```
Comment (freekmurze) [inline src/Events/StatusUpdated.php]: Remove this docblock

File: src/Events/StatusUpdated.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public function getOldStatus(): string
+    {
+        return $this->oldStatus;
+    }
+
+    /**
+     * @return string
```
Comment (freekmurze) [inline src/Events/StatusUpdated.php]: Remove this docblock

File: src/Events/StatusUpdated.php [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+     */
+    public function getNewStatus(): string
+    {
+        return $this->newStatus;
+    }
+
+    /**
+     * @return \Illuminate\Database\Eloquent\Model
```
Comment (freekmurze) [inline src/Events/StatusUpdated.php]: Remove this docblock

File: src/HasStatuses.php [hunk 6]
```diff
@@ -2,38 +2,46 @@
...
 use Illuminate\Database\Eloquent\Relations\MorphMany;
 use Illuminate\Database\Query\Builder as QueryBuilder;
+use Illuminate\Support\Facades\DB;
+use Spatie\ModelStatus\Events\StatusUpdated;
+use Spatie\ModelStatus\Exceptions\InvalidStatus;
 
 trait HasStatuses
 {
+
+    /**
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove this docblock

File: src/HasStatuses.php [hunk 7]
```diff
@@ -2,38 +2,46 @@
...
+     *
+     * @return \Illuminate\Database\Eloquent\Relations\MorphMany
+     */
     public function statuses(): MorphMany
     {
         return $this->morphMany($this->getStatusModelClassName(), 'model')->latest();
     }
 
+    /**
+     * Get the latest model status
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove this docblock

File: src/HasStatuses.php [hunk 8]
```diff
@@ -2,38 +2,46 @@
...
-        $this->statuses()->create([
-            'name' => $name,
-            'reason' => $reason,
-        ]);
-
-        return $this;
-    }
-
+    /**
+     * Allows checking if a status is valid before actually setting it. This function is used by the setStatus
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove this docblock

File: src/HasStatuses.php [hunk 9]
```diff
@@ -46,29 +54,36 @@ public function isValidStatus(string $name, string $reason = ''): bool
      */
     public function latestStatus(...$names): ?Status
     {
-        $names = is_array($names) ? array_flatten($names) : func_get_args();
+        $names = \is_array($names) ? array_flatten($names) : \func_get_args();
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove `/`. We are sure we don't have that function in the namespace.

File: src/HasStatuses.php [hunk 10]
```diff
@@ -46,29 +54,36 @@ public function isValidStatus(string $name, string $reason = ''): bool
...
 
-        if (count($names) < 1) {
+        if (\count($names) < 1)
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove that `/`. We are sure we don't have that function in the namespace.

File: src/HasStatuses.php [hunk 11]
```diff
@@ -46,29 +54,36 @@ public function isValidStatus(string $name, string $reason = ''): bool
...
+        {
             return $this->statuses()->orderByDesc('id')->first();
         }
 
         return $this->statuses()->whereIn('name', $names)->orderByDesc('id')->first();
     }
 
+    /**
+     * @param \Illuminate\Database\Eloquent\Builder $builder
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove the docblock

File: src/HasStatuses.php [hunk 12]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
+                                    ->from($this->getStatusTableName())
+                                    ->where('model_type', static::class)
+                                    ->groupBy('model_id');
+                            });
+                })
             ->orWhereDoesntHave('statuses');
     }
 
+    /**
+     * Easy property access to the latest status
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove this docblock

File: src/HasStatuses.php [hunk 13]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
     public function getStatusAttribute(): string
     {
-        return (string) $this->latestStatus();
+        return (string)$this->latestStatus();
+    }
+
+
+    /**
+     * Change the model status
+     *
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove this docblock

File: src/HasStatuses.php [hunk 14]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
+            throw InvalidStatus::create($name);
+        }
+
+        return $this->forceSetStatus($name, $reason);
     }
 
+    /**
+     * Force a status change without checking the validity of the new status.
+     *
+     * @param string $name
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove this docblock

File: src/HasStatuses.php [hunk 15]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
+     */
     public function forceSetStatus(string $name, string $reason = ''): self
     {
+        // Keep track of the previous status before updating
+        $oldStatus = $this->status;
+
+        // Update current status
         $this->statuses()->create([
-            'name' => $name,
+            'name'   => $name,
```
Comment (freekmurze) [inline src/HasStatuses.php]: Restore the original spacing.

File: src/HasStatuses.php [hunk 16]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
             'reason' => $reason,
         ]);
 
+        // Dispatch an event in case the status has changed
+        if ($oldStatus !== $name)
+        {
+            \event(new StatusUpdated($oldStatus, $name, $this));
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove `/`

File: src/HasStatuses.php [hunk 17]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
+        $oldStatus = $this->status;
+
+        // Update current status
         $this->statuses()->create([
-            'name' => $name,
+            'name'   => $name,
             'reason' => $reason,
         ]);
 
+        // Dispatch an event in case the status has changed
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove the comment

File: src/HasStatuses.php [hunk 18]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
+        if ($oldStatus !== $name)
+        {
+            \event(new StatusUpdated($oldStatus, $name, $this));
+        }
+
         return $this;
     }
 
+    /**
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove the docblock

File: src/HasStatuses.php [hunk 19]
```diff
@@ -78,44 +93,97 @@ public function scopeCurrentStatus(Builder $builder, string $name)
...
+     * @return string
+     */
     protected function getStatusTableName(): string
     {
         $modelClass = $this->getStatusModelClassName();
 
         return (new $modelClass)->getTable();
     }
 
+    /**
```
Comment (freekmurze) [inline src/HasStatuses.php]: Remove the docblock

File: tests/StatusEventsTest.php [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+
+        $this->testModel = TestModel::create([
+            'name' => 'name',
+        ]);
+    }
+
+    /** @test */
+    public function it_fires_an_event_when_status_changes()
+    {
+        // Prepare
```
Comment (freekmurze) [inline tests/StatusEventsTest.php]: Remove all `// Prepare`, `// Act` and `// Assert` comments

File: tests/StatusEventsTest.php [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+        $this->testModel->setStatus('pending', 'waiting on action');
+        Event::fake(); // Fake after, in order not to get the initial event
```
Comment (freekmurze) [inline tests/StatusEventsTest.php]: Remove the comment

File: tests/StatusEventsTest.php [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+                       && $e->getOldStatus() === 'pending';
+            });
+    }
+
+    /** @test */
+    public function it_does_not_fire_an_event_when_status_stays_same()
+    {
+        // Prepare
+        $this->testModel->setStatus('pending', 'waiting on action');
+        Event::fake(); // Fake after, in order not to get the initial event
```
Comment (freekmurze) [inline tests/StatusEventsTest.php]: Remove the comment

File: tests/StatusEventsTest.php [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+
+        // Act
+        $this->testModel->setStatus('status a', 'Reason a');
+
+        // Assert
+        Event::assertDispatched(StatusUpdated::class,
+            function (StatusUpdated $e) {
```
Comment (freekmurze) [inline tests/StatusEventsTest.php]: Rename `$e` to `$event`

File: tests/StatusEventsTest.php [hunk 24]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Spatie\ModelStatus\Tests;
+
+use Illuminate\Support\Facades\Event;
+use Spatie\ModelStatus\Events\StatusUpdated;
+use Spatie\ModelStatus\Tests\Models\TestModel;
+
+class StatusEventsTest extends TestCase
+{
```
Comment (freekmurze) [inline tests/StatusEventsTest.php]: And `/** Spatie\ModelStatus\Tests\Models\TestModel $testModel */`

File: tests/StatusEventsTest.php [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+        Event::assertDispatched(StatusUpdated::class,
+            function (StatusUpdated $e) {
+                return $e->getModel()->is($this->testModel)
+                       && $e->getNewStatus() === 'status a'
+                       && $e->getOldStatus() === 'pending';
+            });
+    }
+
+    /** @test */
+    public function it_does_not_fire_an_event_when_status_stays_same()
```
Comment (freekmurze) [inline tests/StatusEventsTest.php]: Rename to `it_does_not_fire_an_event_when_status_stays_the_same`

File: src/HasStatuses.php (hunk 9 above)
Comment (vpratfr) [reply src/HasStatuses.php]: The `\` added before those functions are suggested by PHPStorm (PHP EA Inspections plugin)

See also: https://github.com/Roave/FunctionFQNReplacer 

If you are aware of that (and pretty sure you probably are), I will remove that too.

**Comment by vpratfr** [comment]
I think I mostly reverted all the changes. Apart from the `\` prefix before a few function names because they are suggested optimisations by PHP Storm plugin PHP EA inspections.

See https://github.com/Roave/FunctionFQNReplacer for benchmark

Let me know if I need to change anything else.

File: src/HasStatuses.php (hunk 9 above)
Comment (kalessil) [reply src/HasStatuses.php]: This (https://github.com/symfony/symfony/commit/bccab35aa2c0a6500281083414346a0417c234ac) commit from Symfony would be more convincing =)

**Comment by freekmurze** [comment]
Remove the remaining `\` before the function names, and after that I think it's fine to merge.

**Comment by vpratfr** [comment]
Done.

Out of curiousity, why do you choose not to follow the advises about the `\` for those functions (70% performance gain for a very small tradeoff) ?

**Comment by freekmurze** [comment]
I'm thinking it's just a micro-optimization.

Thanks for your work on this PR.

**Comment by vpratfr** [comment]
You're welcome. Looking forward to removing that very same code from my own project which needed the events.

I do agree on the "micro" aspect, but when they are free to do, and do not affect anything else than just code style (plus it can just be automated in PHPStorm), then I usually like applying it.

Cheers, have a nice evening.

**Pull request merged**